lua/claimThrottleLock.lua

```diff
@@ -0,0 +1,17 @@
+--
+-- "Claims" a throttle lock by grabbing the key.
+--
+-- KEYS[1]   - key
+-- ARGV[1]   - ttl
+local key     = KEYS[1]
+local ttl = ARGV[1]
+
+local value = redis.call("SET", key, "", "NX", "PX", ttl)
+
+-- If we were the first to set the value, return a sentinel value of 0.
+if value then
+	return 0
```

Comment:
0 is not the best value to return as you might get the same value with PTTL for an existing key. The key might still exist but you get a 0 – this means that the key is going to expire by the next millisecond but at this point of time it is still available. It would be better to return -2 which is the default for "key did not exist".

If you return a 0 then `throttle` function assumes it was able to set a new lock and calls `throttleFn()`. If the key actually existed and you got 0 for an almost-but-not-yet-expired key, then the lock expires by the time you call `throttleFn()` and it would be possible to call the same function immediately again as there is no lock anymore.